### ISSUE TITLE
Add Loan Items - table hard to parse

### ISSUE THREAD

**Issue by ebraker** [ArctosDB/arctos#5564]
I think it is difficult to add items to loans in the new results form since the catalog record #, taxon, and part name are so far away from the checkbox. Here's the old form where the "add" button is very proximal to the data pieces I need to see to be able to grab the right thing:

![image](https://user-images.githubusercontent.com/11336485/215227644-9b4c9383-af89-4327-8c44-99c56ba1a697.png)

...versus  the current form I just held e a ruler to my monitor to make sure I'm selecting the right part for the right specimen:
![image](https://user-images.githubusercontent.com/11336485/215227756-408fbabf-0edc-4b79-94d3-4fd645a9fa5f.png)

Any ideas or can we put this as an agenda item for a UI meeting?

 @campmlc @ccicero have you had issues?

**Comment by Jegelewicz** [comment]
I have never been a fan of the teeny tiny checkboxes....

**Comment by campmlc** [comment]
I haven't used it yet, but based on screenshot I absolutely agree. Is there
a reason why the subsample box is so wide? Can we swap columns?

**Comment by dustymc** [comment]
If the request isn't for like 19 JS libraries trying to play with each other in a sometimes-public form, I'm receptive!

I can put columns in whatever order very easily.

I can make pick boxes narrower, but it'll be harder to see what's what. (I think that's mostly from MSB comboparts so maybe that's a very small price to pay considering the involved data?)

I'm up for WHATEVER as a way to select a part - suggest something better than the checks?

Striping (instead of getting the ruler out) doesn't generally play nice with sortable tables, but I've got a new CSS trick I can try before I insist ya'll pick one.

If there's something obvious it can be done first, then toss whatever's left at a UI meetiing?

**Comment by Jegelewicz** [comment]
> Striping

A good first step!

**Comment by Jegelewicz** [comment]
We can talk check boxes at UI

**Comment by ebraker** [comment]
There are two things that the original 'add item' form does that would be handy to replicate:

1. Clearly distinguish one catalog record from another (see images at top of the issue - catalog records were chunked into separate tables in the old form and super easy to tell apart from the next). The current form doesn't make the change from one catalog number to the next obvious (stripes may help, but the color change would need to correspond to the catnum instead of changing every-other-row, e.g., a record with three parts would need the same stripe color for three rows before changing to a different stripe color). 

2. The button or checkbox should sit next to the part name to make it easy to pick out for records that have multiple parts. My suggested order would be:  catnum, scientificname, part, button/checkbox, subsample, item remark, etc. etc.

**Comment by dustymc** [comment]
> talk check boxes at UI

We should do everything that's possible to do here, where everyone can participate and has plenty of time to consider how this might affect their workflow.

> a record with three parts

Doesn't need "substriping" - top/middle/bottom is easy - but a record with...

```
 select max(c) from (select derived_from_cat_item,count(*) c from specimen_part group by derived_from_cat_item having count(*) > 20) x;
 max 
-----
 200
```

200 parts is another matter.

How sortable do things need to be, and how "columnar" do things need to be? One possibility would involve nested tables, the outer of which would sort by 1:1 (ish, for now) things like GUID and ID, and the inner (which would have different column widths on each record) would sort within a catalog record by part name, condition, etc.

Or if sorting isn't at all important then perhaps I could switch to some sort of CSS rather than tables, which would allow me to keep everything in line (if we can - or should - coerce this into an equal number of columns per record, and perhaps at the expense of 999 records carrying a lot of whitespace to accommodate that one verbose thing).

I'd much rather stick with tables - they are generally more predictable data containers than anything else - but if this demands something tables can't deliver then I'm up for trying something else.

> button or checkbox

Those lead to very different functionality. The button handles one part at a time (because that's all the old form could realistically accommodate). Click the wrong button, and you need to go off to some other form, undo your mistake - perhaps noticed from something 5 records further in - then come back and fix it. The checkbox is part of a "set it all up, review, then save" approach, where you can go correct that previous thing before DOING anything. I can go either way, but I think the "as a batch" approach is probably a lot closer to reflecting what actually happens.

>part, [otherstuff] subsample

Doesn't make sense to me to separate those - this form allows the long-requested 'subsample a skeleton as a rib' functionality, seems to me that the 'source' and 'result' parts should be (very) near each other.

I've also heard that encumbrances are the most important thing here, so probably should be 'leftward'....

And attributes will be the most important thing (175 of those 200 parts might be liver samples, some of them preserved better than others, some with more remaining volume, etc.) so many/most records so those should be nearest the widget....

Do we really need all those containerIDs in there?

If 5 flattened attributes really sufficient? (And I'm not necessarily suggesting flattening more: what if a record has 700 attributes?)

**Comment by campmlc** [comment]
I think this is going to require a separate discussion. I just tried to use the new form to add 13 items to a loan, and could not do so. The new form in its current state is not functional.  At the very least, we need CatNum, Other ID, SciName, Part Disposition, preservation, and Part BARCODE visible as the first columns to the left, followed by subsample y/n and subsample part name. Right now, to find a barcode, I have to scroll so far to the right I see nothing else and lose the select tool. Also, it appears that in order to subsample, currently the user must write in the name of the subsampled part for each selection, rather than just selecting yes/no as we used to, with an option of renaming. I could technically write in part names for a small loan, but that should not be necessary if the subsampled part name is the same as the parent.
We either need a solution that looks and functions like our previous form, or we need some drastic change that will allow the same functionality.

**Comment by cjconroy** [comment]
I am similarly confused. I am trying to add two skulls to a loan. There is only a button that says add all checked parts to loan. I can't do that because one of the parts in the table is a duplicate of one of the skulls from a previous, closed loan. Even after checking the boxes, there does not seem to be a way to add those items to the loan!  What is going on?

**Comment by dustymc** [comment]
This is the separate discussion, and the form is very much functional.

We can't discuss what column order is until we decide if there are columns or not. Please see above comments, I don't think we can progress without that discussion.

>not be necessary if the subsampled part name is the same as the parent

How would I know that? (I could potentially pre-fill the new part name with the old and users can change it as necessary?)

@cjconroy I think maybe you're on the old form? Screenshots would be very helpful.

**Comment by cjconroy** [comment]
looks like this to me
![Uploading Screenshot 2023-01-31 at 9.44.05 AM.png…]()

**Comment by campmlc** [comment]
Not being able to see the barcode and preservation next to the part name
and guid and the select button means, for my purposes, the form is not
usable. Prefilled subsample name that allows edits works.

**Comment by cjconroy** [comment]
OK, my bad complaining about the button tool. However, is there a way to avoid seeing duplicate skulls from previous closed transactions?

**Comment by dustymc** [comment]
make checkboxes bigger somehow?

**Comment by dustymc** [comment]
don't allow checkbox if part is already on loan

**Comment by dustymc** [comment]
prioritize striping over sorting

**Comment by dustymc** [comment]
Are all of these necessary? 

began_date,
ended_date,
verbatim_date,
received_date, (from accession)

Anything else that can be removed or should be added?

**Comment by ccicero** [comment]
I would keep 'verbatim_date' but am fine with removing the others.

What about locality info - higher geog and spec loc? seems important to have that as well.

I'm not sure if I can make the meeting next week but @atrox10 will be there for MVZ (+ MK). My personal preference would be this order:
Add to Loan
CatNum
Scientific Name
Part
Attr1 (preservation)
Disposition
Loan
locality info - ???
VerbatimDate
all the part location fields
then the rest

Can we hide blank columns (e.g., if no values in Attr2-5)?

**Comment by dustymc** [comment]
>locality info

I can add whatever, but that's going to get real complicated real fast so I'll need details - just priority, all, maybe https://github.com/ArctosDB/arctos/issues/5594 (or various other locality attributes), ????

>order

I've been hearing that the old form was liked and sorting is secondary, so I'm rebuilding to follow that idea, with the parts nested in a cell of the record's row.  There are essentially two column sorts, one for 'record stuff' and another for 'part stuff.'

> Attr1 (preservation)

That is not a valid association - attr1 can be most anything in certain situations. I could dedicate columns to types or something, but I think it would end up with a lot more columns, and a lot more whitespace to scroll past for some collections (or even 'subcollections' - tissues vs bones).

>hide blank c

Not easily, and I think without somehow backing up and going in some other direction. (That's definitely not a 'no' but I can't quite wrap my head around how I'd do it at the moment.) I've been trying to keep them somewhat tabular (so they line up with other "subtables") - even letting them stretch to fit data hurt my brain, I think table-like is easier to see. The idea is that importance decreases as you go right, so you can just not scroll over there unless you've got a reason to.

Here's what it looks like at the moment - timely feedback most appreciated.

<img width="1691" alt="Screenshot 2023-02-03 at 10 03 09 AM" src="https://user-images.githubusercontent.com/5720791/216675043-26d7ba19-18b1-4279-866f-c39c4200019c.png">

**Comment by ebraker** [comment]
That looks a lot better already. I think Part name should be the field directly next to the checkbox, then subsample. It sounds like @campmlc needs barcode, so I'd put that somewhere visible (maybe following itemdispn?).

> don't allow checkbox if part is already on loan
This will not work for media loans - we need to be able to add e.g. CT scans to multiple concurrent loans.

**Comment by dustymc** [comment]
checkbox-partname-subsample is fine/reasonable.

I don't think breaking one of the 21 container-columns out all by itself is reasonable. (Do we need all that there? Would plpath be better? plpath and the container-containing-the-part barcode?? Just the one barcode???)

>catnum
sciname
then the Add 

I think @ccicero would put a bunch of locality stuff in there. Maybe I'll see if I can figure out how to let users hide columns without making this thing so complicated that it can't be maintained??

>not work for media loans 

That's more compelling than whatever the other argument was (@AJLinn maybe??). I'll see if I can make the checkbox red or something when the part's already on loan - bueno?

**Comment by campmlc** [comment]
I can't fully respond here yet because I'm dealing with other work crises -
but for my purposes, all I need in the loan request form is the ultimate
container barcode which contains the part. I don't need any of the other
PartsTableDownload path.
I need only guid, scinname, customID, part name, part barcode, part
attributes (these could be JSONified), part disposition, and encumbrances.
It would be great if we could customize the loan pick form the same way we
customize the data entry form, by choosing which fields to display and
which to hide.

**Comment by campmlc** [comment]
Also need a way to subsample y/n and option to change the name of the
subsample (subsample as . . . )
We used to have bulk tools at the top of the page that would allow us to
select subsample all - perhaps bring that back and say subsample all (part
name of x ) as (part name of y).

**Comment by ebraker** [comment]
I'm assuming the order on the old form worked for most people (?). By the time we are adding specific parts (e.g., screen #2 of "add items to loan"), we've already selected the exact catalog records we are working with and used relevant date/locality info on the search result page to guide us, so I sort of think all of that other catalog record metadata can be pushed right. The metadata are still there a scroll away, but the cat num, scientific_name, and parts/barcodes/subsample etc. fields are most relevant by the time you are adding parts. But I could be wrong, so if it is simple to hide columns, great!

![image](https://user-images.githubusercontent.com/11336485/216707204-69e7a1eb-14ed-4348-8a93-5945efb125c3.png)

> make the checkbox red

Works for me

**Comment by dustymc** [comment]
>don't need any of the other
PartsTableDownload path.

Awesome, I'm I'm zapping it.

>bulk tools

You're going to have to be a LOT more specific before I can write code to that (and there's a bulkloader, this is pretty explicitly the not-bulk thing?)

>old form worked for most people

Nah, everybody hated it - until it changed....

I think I've got a workable recipe to hide columns, and if I can zap 21 checkboxes before they exist maybe even a *usable* workable recipe!

**Comment by ebraker** [comment]
The new tool is great! Way more functional. Thank you!

**Comment by dustymc** [comment]
* add filter - only show things with (disposition)
* set default disposition select
*

**Comment by Jegelewicz** [comment]
From UI meeting today

1. Remove the itemDispn column and instead add a method to "change all checked to" and select the disposition
2. Can we have the ability to filter out items that have on loan disposition (or only get in collection) or a comma separated list of things to filter for or out?

**Comment by dustymc** [comment]
>Remove the itemDispn column 

No.

> ability to filter

Anything other than disposition?

**Comment by campmlc** [comment]
Than can we rename ItmDspn as Loan ItmDspn. It is way too confusing.

**Comment by dustymc** [comment]
We can rename any column to whatever you want, here's the place to do it. (But you were complaining about things being pushed off of your screen, and the more verbose the labels the more unavoidable that becomes.)

**Comment by Jegelewicz** [comment]
Why is this a hard No?

> Remove the itemDispn column and instead add a method to "change all checked to" and select the disposition

Isn't that how the old form worked? I don't really do loans, so I don't know for sure, but if it was, it seems like everyone was enjoying that (at least those in the meeting yesterday agreed it was the way to go). I just want to make sure we explore all of the options.

**Comment by campmlc** [comment]
I for one will never use that column. Everything I do will result in a loan
item disposition of on loan. Why should I lose screen real estate to that?

**Comment by ewommack** [comment]
> No, I think it forced everything to 'on loan' and there are several use cases (exhibits, that thing I don't really want to know about involving media) where that's not appropriate or desirable.

I think what we heard about yesterday was that there are a number of collections that use loans to help keep track of use of objects as they move around both digitally and physically. Having things just either go on loan or not was limiting those collections ability to track and use the loan forms, and now with the ability to change to more disposition types the forms is much more versatile for them. 
@AJLinn - did I get that right?

**Comment by ebraker** [comment]
The old form (which is still accessible) did not include the item disposition, only the object disposition, so it was never confusing:


![Image](https://user-images.githubusercontent.com/11336485/218224962-e2a3e249-c700-44d7-90f4-0994597285a1.png)

**Comment by dustymc** [comment]
> did not include the item disposition

... because it forced them all, kicking and screaming, to 'on loan' - a simplicity which Arctos has now outgrown.

**Comment by ebraker** [comment]
was editing my comment as you wrote:

Adding items however, only changed the disposition to "on loan" (I assume users had to manually adjust the disposition for "other" use cases). I agree we need to get rid of that column and just have a "set disposition to [whatever] for all checked parts" up at the top (default could be 'loan' or 'pick'

**Comment by ebraker** [comment]
Now the bulk disposition update functionality is available (thank you!). So can we get rid of the column in the table (it is still confusing) and just make that grey disposition update field yellow up top? e.g.,

![Image](https://user-images.githubusercontent.com/11336485/218337172-1560ce9c-f0fc-42a5-9e0d-d83dd49b7485.png)

I know we can hide columns but I'm still having to scroll to see my part conditions and  part remarks. Removing the Item disposition column and empty space would be really useful, e.g.:

![Image](https://user-images.githubusercontent.com/11336485/218337553-7c6d9c59-0b54-4326-91e7-7cd5e3fdf6c2.png)

Since this is my current view with all the columns possible hidden (still can't see what I need):


![Image](https://user-images.githubusercontent.com/11336485/218337644-76e320f1-d1f9-4a59-9ffd-256c088246f1.png)


If NOT, can we at least have the item disposition value also defaulted to "select disposition" in the table to avoid the confusing "on loan" value?

**Comment by campmlc** [comment]
Agree with @ebraker

**Comment by ewommack** [comment]
@AJLinn said the ItemDisp column was really helpful for her collection. We really need to hear from other collections that use the form too.

**Comment by campmlc** [comment]
Is that still the case if there is a bulk select loan disposition at the
top of the page?

**Comment by dustymc** [comment]
> really need to hear from other collections 

We have heard from there, there was a meeting, the column is necessary.

**Comment by ebraker** [comment]
@ewommack the item disposition is separate from the object disposition, which is what was requested. The functionality of the tool would be present.

Agree with Mariel re:

> Is that still the case if there is a bulk select loan disposition at the
top of the page?

@dustymc sorry but this is not working for me  so I am  reopening this issue. 

1. Can we put the item disposition tool above? Or can you explain why this isn't possible?
![image](https://user-images.githubusercontent.com/11336485/218512039-7a0ca9fa-e4fa-4f83-9b16-df96d6af58b0.png)
2. If not (and I really hope #1 is possible), can the default for this menu be set to  "select disposition" vs "on loan" to avoid confusion?
3. Can we make this table have draggable columns? I need to reorder since I can't see the actual columns I need which are Part condition, Part remarks, and Object disposition -  I need to know if something is not stable enough to go on loan or on loan/used up/deaccessioned/missing to make transaction decisions (and I might not filter out on loan if doing a media loan where simultaneous borrowing is possible).


There's also blank columns taking up space - can this go away (I have two essential columns that I have to scroll to see):
![image](https://user-images.githubusercontent.com/11336485/218512338-6f84734a-6d9f-4a1e-9429-0f955d73c15c.png)

**Comment by dustymc** [comment]
> item disposition is separate from the object disposition,

What does this mean? There is only one disposition, the form shows current and 'will become' states.

>can you explain why this isn't possible

Because various people (Including you! "and I might not filter out on loan if doing a media loan where simultaneous borrowing is possible" ????????????????) have suggested that not all things going on loan need to become 'on loan'

>can the default for this menu be set to "select disposition" vs "on loan" 

I suppose, but it'll be an extra click to 'on loan' for much/most usage.

>draggable columns? 

Not easily. We can, but I think not while reusing any code.

>blank columns taking up space

I need more than a tiny peek, shouldn't be like that.

**Comment by ebraker** [comment]
@dustymc I understand why this field is important and I am not asking to get rid of it. Item disposition and object disposition are separate in this case, because these are two different columns available on this form, both of which take up space (the distinction is object disposition = existing disp and item disp = future disposition/what you are changing your objects to, e.g. putting on loan, exhibit, etc). 

My ask is not to do away with the item disp (change to) column. Ideally it is not part of the table, and instead there is a tool directly at the top of the table (yellow/required), where you are able to batch change clicked records, e.g., "set all items to [selected disposition status]"  so people don't have to click individually. 


> draggable columns?

Not easily. We can, but I think not while reusing any code.

That would be great


> I need more than a tiny peek, shouldn't be like that.

Does this help?

![image](https://user-images.githubusercontent.com/11336485/218522585-ecb45ae7-176b-4742-ab24-838bd6555380.png)

**Comment by dustymc** [comment]
bold labels in docs (top)
possible to highlight on mouseover header (not while reusing code...)
can we stack disposition? (@ebraker had mostly stacked-something stretching lines out so would mostly be using wasted space)
* current disposition
* update to disposition

**Comment by dustymc** [comment]
<img width="482" alt="Screenshot 2023-02-14 at 5 02 47 PM" src="https://user-images.githubusercontent.com/5720791/218898288-9e442d09-d3bc-4ea0-b54b-ee670d1f9765.png">

**Comment by campmlc** [comment]
This looks pretty good? I'll have to try out once back to my computer.


Mariel Campbell   (she/her)
Collection Manager, Division of Genomic Resources
Museum of Southwestern  Biology

Shipping Address:
Museum of Southwestern Biology
Division of Genomic Resources
1 University of New Mexico
MSC03 2020
Albuquerque, NM 87131

(505) 277-7808 office
email: ***@***.***


________________________________
